internal/cmd/rootcmd/signal.go

```diff
@@ -0,0 +1,54 @@
+package rootcmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/bombsimon/logrusr/v2"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+var shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}
+
+var onlyOneSignalHandler = make(chan struct{})
+
+// SetupSignalHandler registers for SIGTERM and SIGINT. A context is returned
+// which is canceled on one of these signals. If a second signal is not caught, the program
+// will delay for the configured period of time before terminating. If a second signal is caught,
+// the program is terminated with exit code 1.
+func SetupSignalHandler(cfg *manager.Config) (context.Context, error) {
+
+	deprecatedLogger, err := util.MakeLogger(cfg.LogLevel, cfg.LogFormat)
+	if err != nil {
+		return nil, err
+	}
+	logger := logrusr.New(deprecatedLogger)
+
+	close(onlyOneSignalHandler) // panics when called twice
```

Comment:
Continued from above: I would like to avoid the panic and instead log a clear error and exit here instead of the current implementation which could produce somewhat confusing output.